File: src/liveplugin/pluginrunner/GroovyPluginRunner.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package liveplugin.pluginrunner;

import com.intellij.util.Function;
import groovy.lang.Binding;
import groovy.util.GroovyScriptEngine;
import org.codehaus.groovy.control.CompilationFailedException;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static liveplugin.MyFileUtil.*;
import static liveplugin.pluginrunner.PluginRunner.ClasspathAddition.*;

public class GroovyPluginRunner implements PluginRunner {
	public static final String MAIN_SCRIPT = "plugin.groovy";
	public static final String TEST_SCRIPT = "plugin-test.groovy";
	public static final String GROOVY_ADD_TO_CLASSPATH_KEYWORD = "// " + ADD_TO_CLASSPATH_KEYWORD;
	public static final String GROOVY_DEPENDS_ON_PLUGIN_KEYWORD = "// " + DEPENDS_ON_PLUGIN_KEYWORD;

	private final String scriptName;
	private final ErrorReporter errorReporter;
	private final Map<String, String> environment;

	public GroovyPluginRunner(String scriptName, ErrorReporter errorReporter, Map<String, String> environment) {
		this.scriptName = scriptName;
		this.errorReporter = errorReporter;
		this.environment = new HashMap<String, String>(environment);
	}

	@Override public boolean canRunPlugin(String pathToPluginFolder) {
		return findScriptFileIn(pathToPluginFolder, scriptName) != null;
	}

	@Override public void runPlugin(String pathToPluginFolder, String pluginId, Map<String, ?> binding,
	                                Function<Runnable, Void> runOnEDTCallback) {
		File mainScript = findScriptFileIn(pathToPluginFolder, scriptName);
		runGroovyScript(asUrl(mainScript), pathToPluginFolder, pluginId, binding, runOnEDTCallback);
	}

	@Override public String scriptName() {
		return scriptName;
	}

	private void runGroovyScript(final String mainScriptUrl, String pathToPluginFolder, final String pluginId,
	                             final Map<String, ?> binding, Function<Runnable, Void> runPluginCallback) {
		try {
			environment.put("PLUGIN_PATH", pathToPluginFolder);

			List<String> dependentPlugins = findPluginDependencies(readLines(mainScriptUrl), GROOVY_DEPENDS_ON_PLUGIN_KEYWORD);
			List<String> pathsToAdd = findClasspathAdditions(readLines(mainScriptUrl), GROOVY_ADD_TO_CLASSPATH_KEYWORD, environment, new Function<String, Void>() {
				@Override public Void fun(String path) {
					errorReporter.addLoadingError(pluginId, "Couldn't find dependency '" + path + "'");
					return null;
				}
			});
			String pluginFolderUrl = "file:///" + pathToPluginFolder + "/"; // prefix with "file:///" so that unix-like path works on windows
			pathsToAdd.add(pluginFolderUrl);
			ClassLoader classLoader = createClassLoaderWithDependencies(pathsToAdd, dependentPlugins, mainScriptUrl, pluginId, errorReporter);

			// assume that GroovyScriptEngine is thread-safe
			// (according to this http://groovy.329449.n5.nabble.com/Is-the-GroovyScriptEngine-thread-safe-td331407.html)
			final GroovyScriptEngine scriptEngine = new GroovyScriptEngine(pluginFolderUrl, classLoader);
			try {
				scriptEngine.loadScriptByName(mainScriptUrl);
			} catch (Exception e) {
				errorReporter.addLoadingError(pluginId, e);
				return;
			}

			runPluginCallback.fun(new Runnable() {
				@Override public void run() {
					try {
						scriptEngine.run(mainScriptUrl, createGroovyBinding(binding));
					} catch (Exception e) {
						errorReporter.addRunningError(pluginId, e);
					}
				}
			});

		} catch (IOException e) {
			errorReporter.addLoadingError(pluginId, "Error creating scripting engine. " + e.getMessage());
		} catch (CompilationFailedException e) {
			errorReporter.addLoadingError(pluginId, "Error compiling script. " + e.getMessage());
		} catch (LinkageError e) {
			errorReporter.addLoadingError(pluginId, "Error linking script. " + e.getMessage());
		} catch (Error e) {
			errorReporter.addLoadingError(pluginId, e);
		} catch (Exception e) {
			errorReporter.addLoadingError(pluginId, e);
		}
	}

    private static Binding createGroovyBinding(Map<String, ?> binding) {
        Binding result = new Binding();
        for (Map.Entry<String, ?> entry : binding.entrySet()) {
            result.setVariable(entry.getKey(), entry.getValue());
        }
        return result;
    }
}

File: src/main/liveplugin/pluginrunner/GroovyPluginRunner.kt
package liveplugin.pluginrunner

import groovy.lang.Binding
import groovy.util.GroovyScriptEngine
import liveplugin.findScriptFileIn
import liveplugin.pluginrunner.PluginRunner.ClasspathAddition.createClassLoaderWithDependencies
import liveplugin.pluginrunner.PluginRunner.ClasspathAddition.findClasspathAdditions
import liveplugin.pluginrunner.PluginRunner.ClasspathAddition.findPluginDependencies
import liveplugin.readLines
import liveplugin.toUrlString
import org.codehaus.groovy.control.CompilationFailedException
import java.io.File
import java.io.IOException

class GroovyPluginRunner(
    private val scriptName: String,
    private val errorReporter: ErrorReporter,
    private val environment: Map<String, String>
): PluginRunner {

    override fun scriptName() = scriptName

    override fun runPlugin(pluginFolderPath: String, pluginId: String, binding: Map<String, *>, runOnEDT: (() -> Unit) -> Unit) {
        val mainScript = findScriptFileIn(pluginFolderPath, scriptName)!!
        runGroovyScript(mainScript.toUrlString(), pluginFolderPath, pluginId, binding, runOnEDT)
    }

    private fun runGroovyScript(
        mainScriptUrl: String,
        pluginFolderPath: String,
        pluginId: String,
        binding: Map<String, *>,
        runOnEDT: (() -> Unit) -> Unit
    ) {
        try {
            val environment = environment + Pair("PLUGIN_PATH", pluginFolderPath)

            val dependentPlugins = findPluginDependencies(readLines(mainScriptUrl), groovyDependsOnPluginKeyword)
            val pathsToAdd = findClasspathAdditions(readLines(mainScriptUrl), groovyAddToClasspathKeyword, environment, onError = { path ->
                errorReporter.addLoadingError(pluginId, "Couldn't find dependency '$path'")
            }).map{ File(it) }.toMutableList()
            val pluginFolderUrl = "file:///$pluginFolderPath/" // prefix with "file:///" so that unix-like path works on windows
            pathsToAdd.add(File(pluginFolderPath))
            val classLoader = createClassLoaderWithDependencies(pathsToAdd, dependentPlugins, pluginId, errorReporter)

            // assume that GroovyScriptEngine is thread-safe
            // (according to this http://groovy.329449.n5.nabble.com/Is-the-GroovyScriptEngine-thread-safe-td331407.html)
            val scriptEngine = GroovyScriptEngine(pluginFolderUrl, classLoader)
            try {
                scriptEngine.loadScriptByName(mainScriptUrl)
            } catch (e: Exception) {
                errorReporter.addLoadingError(pluginId, e)
                return
            }

            runOnEDT {
                try {
                    scriptEngine.run(mainScriptUrl, createGroovyBinding(binding))
                } catch (e: Exception) {
                    errorReporter.addRunningError(pluginId, e)
                }
            }

        } catch (e: IOException) {
            errorReporter.addLoadingError(pluginId, "Error creating scripting engine. ${e.message}")
        } catch (e: CompilationFailedException) {
            errorReporter.addLoadingError(pluginId, "Error compiling script. ${e.message}")
        } catch (e: LinkageError) {
            errorReporter.addLoadingError(pluginId, "Error linking script. ${e.message}")
        } catch (e: Error) {
            errorReporter.addLoadingError(pluginId, e)
        } catch (e: Exception) {
            errorReporter.addLoadingError(pluginId, e)
        }
    }

    companion object {
        const val mainScript = "plugin.groovy"
        const val testScript = "plugin-test.groovy"
        const val groovyAddToClasspathKeyword = "// " + PluginRunner.addToClasspathKeyword
        const val groovyDependsOnPluginKeyword = "// " + PluginRunner.dependsOnPluginKeyword

        private fun createGroovyBinding(binding: Map<String, *>): Binding {
            val result = Binding()
            for ((key, value) in binding) {
                result.setVariable(key, value)
            }
            return result
        }
    }
}
